### PR TITLE
lib-user: Add usePanoptesAuthUser, refactor related hooks and components

### PR DESCRIPTION
**Pull request by mcbouslog**
## Package
- lib-user

## Describe your changes
- add `usePanoptesAuthUser` hook for authorized user
- refactor `usePanoptesUser` hook to return user with `avatar_src` if don't already have user with `avatar_src` (almost adds functionality for an admin to see a different user's stats, that's done in #5980 )
- refactor hooks that make a request that uses authorization
- refactor to `useStats` hook per the source of stats expected (`sourceId`), generalizes hook for user or user group

## Context
Previously, lib-user hooks were inconsistently using the authClient passed as prop or the panoptes-javascript-client auth specifically. This created issues when running the dev server for local development and otherwise. These PR refactors make lib-user truly agnostic about the authClient (auth, oauth, other) passed to the components. 

Some related additional potential refactors could be to add React context for the authClient and authorized user (`authUser`), and/or to refactor lib-react-components `usePanoptesUser` to be authClient agnostic, though I don't think either of those are warranted at the moment. Alternatively, there's a fair amount of added complexity for lib-user to be authClient agnostic, maybe that should be revisited after launch and app-root is more established, in conjunction with larger FEM apps/library organization - not a problem for the short term, but noting for consideration in the long term.

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX? n/a
- What user actions should my reviewer step through to review this PR?
  - review https://local.zooniverse.org:8080/?users=markb-panoptes/groups (replace markb-panoptes with your staging user login)
  - review https://local.zooniverse.org:8080/?users=markb-panoptes/stats (replace markb-panoptes with your staging user login)
  - review https://local.zooniverse.org:8080/?groups=1326945 (replace 1326945 with a user group you belong to)
- Which storybook stories should be reviewed? n/a
- Are there plans for follow up PR’s to further fix this bug or develop this feature? yes, see linked GitHub project, but specifically:
  - pass login in url param from app-root to lib-user components 
  - add containers to main components for cleaner code, can create stories and tests for components 
  - pass adminMode boolean from app-root to lib-user, update usePanoptesUser and usePanoptesUserGroup hooks accordingly
  - add sad paths (no authUser, no stats source, errors, loading, etc.) to containers
  - refactor GroupCards as links (with tab focus)

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## Refactoring
- [ ] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

### CODE REVIEW

**Comment by mcbouslog** [comment]
Changing this to draft, as I might review and refactor auth usage in this PR, though if that gets messy I'll do the additional auth refactor in mind in a separate PR.

**Review by goplayoutside3 — COMMENTED** [review]
I am able to run app-root and see my user's stats + groups. However, when I run lib-user in a new (signed-out) browser window, and click the Login button, there's an error setting the `appId`.
Screenshot 2024-03-19 at 3 00 10 PM

Similar to https://github.com/zooniverse/front-end-monorepo/pull/5980, lib-user still hangs in my browser so I thought I'd try a new private window and different browsers. Are you able to sign out and sign in again to lib-user when running locally?

**Comment by mcbouslog** [comment]
> Similar to #5980, lib-user still hangs in my browser so I thought I'd try a new private window and different browsers. Are you able to sign out and sign in again to lib-user when running locally?

I refactored how the app initializes oauth, does that help? I'm able to sign in and out in a Chrome and Firefox window, and after the change I just pushed, also in an incognito Chrome window. I think the load event, which was calling the oauth init, was running before the React component first render, then when the React component did render it never ran `initUserAuth`.

File: packages/lib-user/src/hooks/usePanoptesMemberships.js [hunk 1]
```diff
@@ -11,25 +13,19 @@ const SWRoptions = {
...
-  // logged in
-  if (userID && authorization) {
+async function fetchMemberships({ query, authorization }) {
+  try {
     const { body } = await panoptes.get(endpoint, query, { authorization })
-    return body || {}
+    return body
+  } catch (error) {
+    console.log(error)
+    return null
```
[the remark this answers is not in the document]
Comment (mcbouslog) [reply packages/lib-user/src/hooks/usePanoptesMemberships.js:22]: I hadn't thought about it, that's a great question!

File: packages/lib-user/src/components/UserStats/UserStats.js [hunk 2]
```diff
@@ -20,28 +21,66 @@ import {
...
+  const {
+    data: user,
+    error: userError,
+    isLoading: userLoading
+  } = usePanoptesUser({
+    authClient,
+    authUser,
+    authUserId: authUser?.id,
+    login: authUser?.login // will be changed per app-root login param in subsequent PR
+  })
```
[the remark this answers is not in the document]
Comment (mcbouslog) [reply packages/lib-user/src/components/UserStats/UserStats.js:49]: Yes, definitely should have a readme with the main components (UserStats, GroupStats) and hooks. I'll start working on one for UserStats today, then the hooks. I should have GroupStats finished soon, and will create one for that component last.

Regardless, I'll add more in-line comments going forward.

**Review by goplayoutside3 — APPROVED** [review]
Thanks for the follow-up changes! I can now run lib-user locally and sign-in or sign-out, see my stats and groups, etc. A couple of non-blocking questions:

1. I noticed during review that when I'm viewing a page with a stats chart, on page refresh the chart always defaults back to 7 Days as the date range. This happens in lib-user and app-root. I think it's probably intentional, but just curious if you plan to leave `dataRange` as a variable relevant to only BarChart and it's corresponding dropdown, or have you or Sean considered UX where date range is a url param too?

2. Just a suggestion to include some documentation about why you're making lib-user components authClient agnostic - either as a detailed Github comment in this PR, or a future Readme.

**Comment by mcbouslog** [comment]
> 1. I noticed during review that when I'm viewing a page with a stats chart, on page refresh the chart always defaults back to 7 Days as the date range. This happens in lib-user and app-root. I think it's probably intentional, but just curious if you plan to leave `dataRange` as a variable relevant to only BarChart and it's corresponding dropdown, or have you or Sean considered UX where date range is a url param too?

Yes! Adding UX for a custom date range and adding the date range as a param are intended before launch, though I'm not sure if they'll be ready for initial core functionality testing (included in "Full v1" column in linked GitHub project)

> 2. Just a suggestion to include some documentation about why you're making lib-user components authClient agnostic - either as a detailed Github comment in this PR, or a future Readme.

Yes, working on some related documentation. And it might be therapeutic personally - if I write the scary stories down they won't scare me as much 🧟 🤣

**Pull request merged**